File: internal/importer/turn.go
package importer

import "github.com/krowkcom/cli/internal/store"

// TurnCandidate is one transcript line reduced to the four things the turn
// rule cares about. It is provider-neutral on purpose: the rule has to give
// the same answer for a Claude JSONL line, a cursor row and an opencode
// message, and it can only do that if each source translates into this shape
// rather than the rule learning three formats.
type TurnCandidate struct {
	// Role is the message role as the source reported it.
	Role store.Role
	// Meta is the source's own "this was not typed by a person" marker:
	// Claude's isMeta, a hook's injected line, a system reminder. Sources
	// set it from whatever their format calls it.
	Meta bool
	// Attachment is a line that exists to carry a file or a pasted blob
	// into the context — Claude's `type: "attachment"` lines. It wears the
	// user role and is not a prompt.
	Attachment bool
	// Interrupt is a cancelled or interrupted request. It ends work rather
	// than starting it, so it must not open a turn that then swallows the
	// next real prompt.
	Interrupt bool
	// PartTypes is the canonical types of the message's parts, in order.
	// The rule reads it to tell a prompt from a bag of tool results.
	PartTypes []string
}

// StartsTurn reports whether a candidate opens a new turn.
//
// The rule is "a turn starts at a user message and ends before the next
// one", with the whole difficulty in what counts as a user message. The
// transcripts are full of user-role lines a person never typed: hook output,
// attached files, system reminders, and above all tool results, which the
// protocol sends back under the user role because that is where the API puts
// them. Counting those would report ten turns where somebody asked one
// question, and every per-turn cost figure downstream would be wrong by the
// same factor.
//
// So a turn opens on a user-role line that is not marked meta, is not an
// attachment, is not an interrupt, and carries at least one part that is not
// a tool result. The last clause is the one that does the real work: a
// tool_result-only user line is the protocol talking, not a person.
func StartsTurn(c TurnCandidate) bool {
	if c.Role != store.RoleUser {
		return false
	}
	if c.Meta || c.Attachment || c.Interrupt {
		return false
	}
	return hasPromptPart(c.PartTypes)
}

// hasPromptPart reports whether any part is something a person could have
// sent. An empty part list is not a prompt: a user line with no content is
// an artefact, not a question.
func hasPromptPart(types []string) bool {
	for _, t := range types {
		if t != PartToolResult {
			return true
		}
	}
	return false
}

// TurnSpan is the half-open range of message indexes belonging to one turn:
// messages Start through End-1.
type TurnSpan struct {
	Start int
	End   int
}

// SplitTurns groups messages into turns by StartsTurn, covering every
// message exactly once.
//
// Lines before the first prompt get their own leading span. Transcripts do
// begin that way — a resumed session, a system preamble, a hook that ran
// before anybody typed — and the alternatives are both worse: dropping them
// loses transcript, and folding them into the first real turn attributes
// somebody else's work to that prompt. A caller that wants them merged can
// see the leading span for what it is, because it does not start at a
// prompt.
//
// The spans are positional, which is what the store wants: turn i is seq i,
// so a re-read of a growing transcript extends the list instead of
// renumbering it — as long as the source re-reads from the start of the
// session, which is why turns are cumulative and messages are not.
func SplitTurns(candidates []TurnCandidate) []TurnSpan {
	var spans []TurnSpan
	for i, c := range candidates {
		// Index zero always opens a span, prompt or not: that is the
		// leading span, and without it the messages before the first
		// prompt would belong to nothing.
		if i != 0 && !StartsTurn(c) {
			continue
		}
		if n := len(spans); n > 0 {
			spans[n-1].End = i
		}
		spans = append(spans, TurnSpan{Start: i, End: len(candidates)})
	}
	return spans
}

File: crates/krowk-import/src/turn.rs
//! Where one turn ends and the next begins: a user prompt starts a turn;
//! tool results, meta lines, attachments and interrupts do not.

use krowk_store::Role;

#[derive(Debug, Clone, Default)]
pub struct TurnCandidate {
    pub role: Option<Role>,
    pub meta: bool,
    pub attachment: bool,
    pub interrupt: bool,
    pub part_types: Vec<String>,
}

pub fn starts_turn(c: &TurnCandidate) -> bool {
    c.role == Some(Role::User) && !c.meta && !c.attachment && !c.interrupt && c.part_types.iter().any(|t| t != crate::PART_TOOL_RESULT)
}

/// Half-open [start, end) index spans over the candidates. The first always
/// starts a turn, whatever it is, so nothing lands outside one.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct TurnSpan {
    pub start: usize,
    pub end: usize,
}

pub fn split_turns(candidates: &[TurnCandidate]) -> Vec<TurnSpan> {
    let mut spans: Vec<TurnSpan> = Vec::new();
    for (i, c) in candidates.iter().enumerate() {
        if i != 0 && !starts_turn(c) {
            continue;
        }
        if let Some(last) = spans.last_mut() {
            last.end = i;
        }
        spans.push(TurnSpan { start: i, end: candidates.len() });
    }
    spans
}

/// Names each message's turn, so the store links the message to its turn
/// row: message i is candidate i, and belongs to the span holding it.
pub fn link_turns(messages: &mut [krowk_store::Message], spans: &[TurnSpan]) {
    for (seq, span) in spans.iter().enumerate() {
        for m in messages.iter_mut().take(span.end).skip(span.start) {
            m.turn_seq = Some(seq as i64);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn prompts_start_turns_and_tool_results_do_not() {
        let user = |types: &[&str]| TurnCandidate { role: Some(Role::User), part_types: types.iter().map(|t| t.to_string()).collect(), ..Default::default() };
        let asst = TurnCandidate { role: Some(Role::Assistant), ..Default::default() };
        let spans = split_turns(&[user(&["text"]), asst.clone(), user(&["tool_result"]), asst.clone(), user(&["text"]), asst]);
        assert_eq!(spans, vec![TurnSpan { start: 0, end: 4 }, TurnSpan { start: 4, end: 6 }]);
        assert!(!starts_turn(&TurnCandidate { meta: true, ..user(&["text"]) }));
    }
}
